vavr/src/main/java/io/vavr/collection/Traversable.java

```diff
@@ -212,23 +212,25 @@
         if (isEmpty()) {
             return Option.none();
         } else {
-            final Traversable<?> objects = isTraversableAgain() ? this : toStream();
-            final Object o = objects.head();
-            if (o instanceof Number) {
-                final Traversable<Number> numbers = (Traversable<Number>) objects;
-                final double d;
-                if (o instanceof Integer || o instanceof Long || o instanceof Byte || o instanceof BigInteger || o instanceof Short) {
-                    d = numbers.toJavaStream()
-                            .mapToLong(Number::longValue)
-                            .average()
-                            .getAsDouble();
+            final T head;
+            final java.util.stream.Stream<T> elements;
+            if (isTraversableAgain()) {
+                head = head();
+                elements = toJavaStream();
+            } else {
+                final Iterator<T> iterator = iterator();
+                head = iterator.next();
+                elements = Iterator.of(head).concat(iterator).toJavaStream();
+            }
+            if (head instanceof Number) {
+                final java.util.stream.Stream<Number> numbers = (java.util.stream.Stream<Number>) elements;
+                final double avg;
+                if (head instanceof Integer || head instanceof Long || head instanceof Byte || head instanceof BigInteger || head instanceof Short) {
+                    avg = numbers.mapToLong(Number::longValue).average().getAsDouble();
                 } else {
-                    d = numbers.toJavaStream()
-                            .mapToDouble(Number::doubleValue)
-                            .average()
-                            .getAsDouble();
+                    avg = numbers.mapToDouble(Number::doubleValue).average().getAsDouble();
                 }
-                return Option.some(d);
+                return Option.some(avg);
             } else {
                 throw new UnsupportedOperationException("not numeric");
```

Comment:
Better if to check for this condition first (assert precondition) by inverting the predicate `if(!(head instanceof Number)) throw new UnsupportedOperationException)`

Then, you can remove the `else` altogether and reduce complexity and indentation